client/lib/url-search/use-url-search.ts

```diff
@@ -0,0 +1,79 @@
+import { useState } from '@wordpress/element';
+import { addQueryArgs } from '@wordpress/url';
+import debugFactory from 'debug';
+import page from 'page';
+
+const debug = debugFactory( 'calypso:url-search' );
+
+/**
+ * Function for constructing the url to page to. Here are some examples:
+ * 1. { uri: 'google.com', search:'hello' } --> 'google.com?s=hello'
+ * 2. {
+ *     uri: 'wordpress.com/read/search?q=reader+is+awesome',
+ *     search: 'reader is super awesome'
+ *     queryKey: 'q',
+ *    } --> 'wordpress.com/read/search?q=reader+is+super+awesome'
+ *
+ * @param {object} options the options object
+ * @param {string} options.uri the base uri to modify and add a query to
+ * @param {string} options.search the search term
+ * @param {string} [options.queryKey = s] the key to place in the url.  defaults to s
+ * @returns {string} The built search url
+ */
+export const buildSearchUrl = ( {
+	uri,
+	search,
+	queryKey = 's',
+}: {
+	uri: string;
+	search?: string;
+	queryKey?: string;
+} ) => {
+	const query = addQueryArgs( uri, { [ queryKey ]: search } );
+
+	if ( ! search ) {
+		const queryStringIndex = uri.indexOf( '?' );
+
+		if ( queryStringIndex !== -1 ) {
+			return uri.substring( 0, queryStringIndex );
+		}
+
+		return uri;
+	}
+
+	return query.replace( /%20/g, '+' );
+};
+
+function useUrlSearch( queryKey = 's' ) {
+	const [ isSearchOpen, setSearchOpen ] = useState( false );
+
+	function doSearch( query: string ) {
+		setSearchOpen( '' !== query );
+
+		const searchURL = buildSearchUrl( {
+			uri: window.location.href,
+			search: query,
+			queryKey: queryKey,
+		} );
+
+		debug( 'search for: %s', query );
+		if ( queryKey && query ) {
```

Comment:
We should check if the value that exists on the current url is the same as the one received by `doSearch` in that scenario we don't want to trigger a page change.

There is a bug right now where the page is infinitely reloading this is happening because the `Search` component calls doSearch when it renders for the first time, since this calls `doSearch` again it triggers an infinite loop.

Reply:
Thanks for testing it, [fixed it here]([Fixed buildSearchUrl to return only the path](https://github.com/Automattic/wp-calypso/pull/63352/commits/7fcd2df945625f99359a53cbe4f707808c182ca9)). I noted we can't return the entire URL, just the path so that the page won't reload infinitely